tools/build_defs/framework.bzl

```diff
@@ -309,12 +317,21 @@ def cc_external_rule_impl(ctx, attrs):
     # we need this fictive file in the root to get the path of the root in the script
     empty = fictive_file_in_genroot(ctx.actions, ctx.label.name)
 
+    data_dependencies = ctx.attr.data
+
     define_variables = [
         set_cc_envs,
         "export EXT_BUILD_ROOT=##pwd##",
         "export BUILD_TMPDIR=##tmpdir##",
         "export EXT_BUILD_DEPS=##tmpdir##",
         "export INSTALLDIR=$$EXT_BUILD_ROOT$$/" + empty.file.dirname + "/" + lib_name,
+    ] + [
+        "export {key}={value}".format(
+            key=key,
+            # Prepend the exec root to each $(execpath ) lookup because the working directory will not be the exec root.
+            value=ctx.expand_location(value.replace("$(execpath ", "$$EXT_BUILD_ROOT$$/$(execpath "), data_dependencies)
+        )
```

Comment:
The potential issues I can see with this is:

1.  by encoding the full path we are reducing the chances of getting reproducible builds which reduces the effectiveness of caching for downstream targets - but this is a general problem with `rules_foreign_cc` rules so I'm not expecting this to actually be resolved here but:
2.  these are `data` deps being expanded and so these should not be executed as part of the build as they will be for `target` not `exec` configurations and as such all that can be done with such env variables that point at executable code is for the path to be baked into outputs and as such reduce the reproducibility as described in `1.`
3. We should also expand targets from `ctx.attr.tools_deps` and `ctx.attr.additional_tools` (as an aside: why are there two attributes for build tools? - and why is it not called `ctx.attr.tools` as per other rules?) as these are the dependencies that are feasible to be ran as part of the build process. 

Having said the above this is definitely heading in the right direction IMO and am happy with this if the target list is updated to add the tools attributes too.

Reply:
> The potential issues I can see with this is:
> 
> 1. by encoding the full path we are reducing the chances of getting reproducible builds which reduces the effectiveness of caching for downstream targets - but this is a general problem with `rules_foreign_cc` rules so I'm not expecting this to actually be resolved here but:

👍 

> 3. We should also expand targets from `ctx.attr.tools_deps` and `ctx.attr.additional_tools`

Done

> (as an aside: why are there two attributes for build tools? - and why is it not called `ctx.attr.tools` as per other rules?) [...]  as these are the dependencies that are feasible to be ran as part of the build process.

I'm not sure... `additional_tools` doesn't transition away from the target configuration, though... (Add `tools_deps` is currently `host` rather than `exec`, which we should probably clean up at some point too :))

> Having said the above this is definitely heading in the right direction IMO and am happy with this if the target list is updated to add the tools attributes too.

Done - thanks!